ipcache: Fix IPcache leak of remote-node IP addresses

pchaigno opened the pull request:
The first commit refactors the code a bit to avoid any refactoring changes in the second commit. The second commit fixes the IPcache leak for remote-node IP addresses. The third commit increases the log level on the error from debug to error to ensure we don't miss such bugs in the future. See commits for details.

cc @joestringer, as the author of https://github.com/cilium/cilium/pull/19765:
Note I fixed the bug by changing the deletion logic to follow the `a.b.c.d/32` format, but maybe I should have changed the insert logic to follow the format `a.b.c.d` instead.

As far as I can see, this bug wasn't released anywhere, hence the `release-note/misc` label.

Fixes: https://github.com/cilium/cilium/pull/19765
Fixes: https://github.com/cilium/cilium/pull/4810
Fixes: https://github.com/cilium/cilium/issues/21844

pchaigno committed:
node: Fix IPcache leak of remote-node IP addresses

Commit 454df0c06 changed how remote-node IP addresses are upserted in
the IPcache. They used to be upserted as e.g. a.b.c.d and with that
commit are upserted as a.b.c.d/32. The deletion logic was however not
updated to follow the same format.

As a result, we can see the following debug logs for IP addresses of
deleted remote nodes:

    msg="Upserting IP into ipcache layer" identity="{remote-node custom-resource false}" ipAddr=10.164.0.29/32 key=0
    [...]
    msg="Attempt to remove non-existing IP from ipcache layer" ipAddr=10.164.0.29

The deletion logic therefore fails and the remote-node IP addresses are
left in the IPcache.

This commit fixes it by also updating the deletion logic to follow the
format a.b.c.d/32.

This regression was caught by the existing unit test but commit
454df0c06 also updated the test to have the incorrect output.

Fixes: 454df0c06 ("ipcache/md: Convert metadata key to netip.Prefix")
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
ipcache: Warn when trying to remove non-existing entry

When removing IPcache entries (e.g., on remote-node deletion), we check
if the entry is even know to us. If it isn't, we increase an error
counter and log... a debug message.

As far as I can see, this should be a warning message instead. The
debug message was introduced in commit c414ebe7; that commit doesn't
provide an explanation for the lower severity.

This commit changes the log level to warn to raise attention to such
issues in the future.

Fixes: c414ebe7 ("ipcache: Fix handling of endpoint IP events")
Signed-off-by: Paul Chaignon <paul@cilium.io>

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -511,7 +511,7 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
 	cachedIdentity, found := ipc.ipToIdentityCache[ip]
 	if !found {
-		scopedLog.Debug("Attempt to remove non-existing IP from ipcache layer")
+		scopedLog.Error("Attempt to remove non-existing IP from ipcache layer")
```
I'm a little concerned that this will cause noisy logs in some scenario, but I'm OK to try this out and see what reports come through up until 1.13 final.

pchaigno replied:
I have the same concern, but we fail on any error in CI, so hopefully we should catch at least the most obvious scenarios.

pchaigno replied:
Well, CI is currently failing because of that error log. See, it's working :sweat_smile:

joestringer reviewed (approved):
Thanks for the fix. This matches the insertion side, so LGTM.

pchaigno committed:
node: Refactor around deleteIPCache to prepare for fix

The subsequent commit will need to change the format of IPs passed to
deleteIPCache (from a.b.c.d to a.b.c.d/32) for some cases. To enable
that, deleteIPCache needs to take a slice of strings instead of a slice
of net.IPs as argument.

This commit implements that refactoring. It doesn't contains any
functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

julianwiedmann commented on pkg/ipcache/ipcache.go:
```diff
@@ -507,7 +507,7 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
 	cachedIdentity, found := ipc.ipToIdentityCache[ip]
 	if !found {
-		scopedLog.Debug("Attempt to remove non-existing IP from ipcache layer")
+		scopedLog.Warn("Attempt to remove non-existing IP from ipcache layer")
```
FYI - caught this in an (unrelated) [draft PR](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.26-kernel-net-next/138/testReport/junit/Suite-k8s-1/26/K8sDatapathConfig_AutoDirectNodeRoutes_Check_direct_connectivity_with_per_endpoint_routes/).
